File: python/6_zigzagConversion.py
def convert(s: str, numRows: int) -> str:
    """
        P     I     N
        A   L S   I G
        Y A   H R
        P     I
    """
    if (len(s) <= 1 or numRows <= 1):
        return s
    res =""
    i = 0
    j = 0
    size = 2*numRows-2
    while(i<numRows):        
        j = i
        while(j<len(s)):
            # deal with | element
            res += s[j]
            # deal with / element and ignore top and bottom
            if (j+size - 2*i < len(s) and (i!=numRows-1 and i!=0)):
                res+=s[j+size - 2*i]            
            j+= size
        i+=1
    return res

if __name__ == "__main__":
    s = "AB"
    n = 1
    print(convert(s, 1))

File: cpp/6_zigzagConversion.cpp
#include <bits/stdc++.h>
using namespace std;
string convert(string s, int numRows)
{
    /*
    PAYPALISHIRING => numRows:4
    P     I    N
    A   L S  I G
    Y A   H R
    P     I


    0           2n-2
    1         . 2n-1
    .     N+1     .
    N-2 N       3n-4 3n-2
    N-1         3n-3

    Pattern : |/|/|/|
    */

    string res;

    if (numRows <=1){
        return s;
    }
    int size= 2*numRows-2;    
    for (int i=0; i<numRows; i++)
    {
        for (int j=i; j<s.length(); j+=size)
        {
            // deal with | col element
            res+= s[j];

            // deal with / element and ignore top and bottom
            if ((i !=0 && i!=numRows-1) && j+size-2*i < s.length())
            {
                res += s[j+size-2*i];
            }
        }

    }
    return res;
}

int main(){
    string s= "abcdefg";
    int nrows = 4;
    string res = convert(s, nrows);
    cout << res << endl;
    return 0;
}

static const int _ = []() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    return 0;
}();
